### ISSUE TITLE
RFC: fix semi-naive and table problems

### ISSUE THREAD

**Issue by oflatt** [egraphs-good/egglog#228]
Right now, we suffer from several related awkward problems.
The biggest is #142, where users of egglog can write rules where things in the right-hand side of a rule might change.

However, there's another problem. In this example, we see that users can write rules that access tables with no default, causing a rule execute only half of it's action:
https://egraphs-good.github.io/egglog/?program=XQAAgADQAAAAAAAAAAAUGQgnfMUD9dO10C6gYDgH7EMCDsP59lqRF2etuT_we0rC4-zdrRMdtQlgkwnRHou9aMqzrfUygj6rjTtUT_55lzsMcxqoclfQM7aYX12UDnF9ACgHhH7nPTjYxz3Y8rMOJ0YuOAqy89w9KfjT-1l2WN_734IA

I propose a simple fix: allow creating new *datatypes* in the actions of the rules. Also allow using `set` to set tables using these datatypes. However, do not allow looking up a value from a table.

See #223 for the difference between datatypes and tables

**Comment by oflatt** [comment]
For further motivation, consider that these "look-ups" in tables on the right hand side of rules must be justified, complicating the proof system significantly.

**Comment by Alex-Fischman** [comment]
I expected to get an error saying that "there's no value for (Num 1) in mayfail", and stopping execution when I ran that example. Why is this not what happened?

**Comment by oflatt** [comment]
Hmm is that what happens on my PR? Either way it's bad

**Comment by Alex-Fischman** [comment]
I don't know what happens on your PR, I'm saying that this is the behavior I expect as a user.

**Comment by saulshanabrook** [comment]
xposting the original discussion I think that led to this issue: https://github.com/egraphs-good/egglog/pull/223#discussion_r1320605325

Thanks for opening a dedicated issue for it!

**Comment by Alex-Fischman** [comment]
Oliver, wasn't this also your first thought in the other discussion? The correct thing to do, if the user does a mutable lookup on a table with no default action, and there is no output for the inputs they gave, is to throw an error. Also, we can't just restrict users to only put non-datatype lookups in the query, because of the :default option.

**Comment by oflatt** [comment]
I think throwing an error is certainly better than what we do now.
However we could just prevent this problem with this proposal. I also realized, as you said, that this proposal necessitates getting rid of the :default option!

I think that if you want a default you can always use a helper rule.

Another idea would be to change the semantics of default so that it takes the default value in queries without being in the database if that makes any sense.

**Comment by Alex-Fischman** [comment]
Can you elaborate on the second paragraph?

I do like the idea of just getting rid of the default option, but you should look at the current list of examples and see which ones use it before making that decision.

(Also, while we're thinking of getting rid of option, as far as I can tell none of the examples use :on_merge.)

**Comment by oflatt** [comment]
I agree, we need a list of examples. This proposal's implementation should come in the form of a PR with nice examples.

For now, here's how I've implemented defaults in the past:
```

(function myfunc (Math) i64)

(rule ((Num a))
        ((set (myfunc (Num a)) my-default))
        :ruleset my-add-defaults-ruleset)
```

And there are two awkward things about this example, which I think we should consider desugaring away?
1. We don't have a way to match all Math expressions
2. We have to run this ruleset in between all our other rulesets

`on_merge` is only used in my terms encoding on my branch right now- we should also consider removing it somehow

**Comment by oflatt** [comment]
Upon further consideration, it sounds like the current plan is to:
1. For now, make it a run-time error when an access fails. That way no actions complete halfway without us knowing.
2. In the long term, the plan is to do the proper demand transformation in this paper: https://drops.dagstuhl.de/opus/volltexte/2022/16235/pdf/LIPIcs-ECOOP-2022-7.pdf

**Comment by yihozhang** [comment]
The proposal is implemented in #461
